### PR TITLE
MeshImporter/XFormImporter refactor (USDU-3 part 3)

### PR DESCRIPTION
**Pull request by judubu**
## Purpose of this PR

**Ticket/Jira #:**
https://jira.unity3d.com/browse/USDU-3
https://jira.unity3d.com/browse/USDU-10

USD geometric Primvars can be given different kind of [interpolation ](https://graphics.pixar.com/usd/docs/api/class_usd_geom_primvar.html#Usd_InterpolationVals) across the uv space of a surface primitive:
- **constant**: one value per-prim
- **uniform**: one value per-face
- **vertex/varying**: one value per-vertex
- **facevarying**: one value per-vertex per-face

Each primvar can also be individually indexed for data-compression purpose.

This is an issue for Unity which requires all attributes of a mesh to be either interpolated per vertex or per-vertex per-face. 
As a consequence it forces us to sanitize the data on the way in before being pushed to GameObjects.

The sanitization was previously done in the MeshImporter right before setting the Unity mesh buffers which was single threaded and made it impossible to properly cross check all attributes arrays for their interpolation value.

This PR introduces a way to perform sanitization of the Mesh and XForm data right after deserialization while using the job system to lessen the impact of interpolation conversion. For that a new ISanitizable interface and 2 new classes SanitizedMeshSample and XFormMeshSample have been added and the jobified ReadAllJob struct have been modified to provide sanitization service after deserialization. 

The sanitization code takes care of:
    - xform conversion
    - triangulation
    - de-indexation of attributes
    - attribute scope conversion

As a consequence the MeshImporter has been greatly simplified.

The SkeletonImporter has been changed and the JointWeights and JointIndices are now deserialized/loaded by the MeshImporter to avoid aan extra mesh deserialization/sanitization.

## Testing

**Functional Testing status:**

The sanitization service has almost 100% coverage but not the MeshImporter itself.

@cguer You should check the import with all the regular suspect (kitchen, human female, [walking teapot](https://pixar-community-production.s3.us-west-1.amazonaws.com/Siggraph/2020/AR_Walking_Teapot_2020_Straight_Line.usdz) ) and do a before after. 

**Performance Testing status:**

I quickly profile the changes and they are a little bit slower both for static and animated data. Although GC has been greatly reduced. I'll update with the actual numbers after my PTO.

## Overall Product Risks

**Complexity:**

High

**Halo Effect:**

High

## Additional information

**Note to reviewers:**

Check the SanitizedMeshSample for the actual sanitization code and the MeshImporter to see how the sanitized data is pushed to Unity Meshes.
There are some minor changes to the deserialization process to flag colors as actual primvars which give us access to the indices and interpolation values.
Check the test usda file to get a feel for the kind of crappy edge case that can come up and if I forgot anything.

**Reminder:**

- [ ] Add entry in CHANGELOG.md _(if applicable)_

### CODE REVIEW

**Review by clusty — APPROVED** [review]
Only comment was around need for both: MeshSample and SanitizedMeshSample & friends classes.

File: package/com.unity.formats.usd/Dependencies/USD.NET.Unity/Geometry/MeshSampleBase.cs [hunk 1]
```diff
@@ -37,21 +36,12 @@ public class MeshSampleBase : PointBasedSample
         /// When not explicitly specified by the shader, "st" should be considered the default uv set.
         /// </summary>
         /// <remarks>
-        /// UV object types should be Vector{2,3,4}[], List of Vector{2,3,4}, or null.
+        /// UV object types should be Vector{2,3}[], List of Vector{2,3}, or null.
```
Comment (jcowles) [inline package/com.unity.formats.usd/Dependencies/USD.NET.Unity/Geometry/MeshSampleBase.cs:39]: Why just 2,3 components?

File: package/com.unity.formats.usd/Dependencies/USD.NET/serialization/Primvar.cs [hunk 2]
```diff
@@ -142,6 +142,10 @@ public Type GetValueType()
             return typeof(T);
         }
 
+        public bool IsArray => typeof(T).IsArray;
+
+        public int Length => IsArray && value != null ? (value as Array).Length : 0;
```
Comment (jcowles) [inline package/com.unity.formats.usd/Dependencies/USD.NET/serialization/Primvar.cs]: Nit pick, but I would add parens around (IsArray && value != null) for readability.

File: package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/CubeImporter.cs [hunk 3]
```diff
@@ -78,7 +78,7 @@ public static class CubeImporter
                 if (usdCube.colors.Length == 1)
                 {
                     // Constant color can just be set on the material.
-                    mat = options.materialMap.InstantiateSolidColor(usdCube.colors[0].gamma);
+                    mat = options.materialMap.InstantiateSolidColor(usdCube.colors.value[0].gamma);
                     Debug.Log("constant colors assigned");
```
Comment (jcowles) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/CubeImporter.cs]: I think we should remove this debug.log

File: package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs [hunk 4]
```diff
@@ -464,44 +483,39 @@ public static GeometrySubsets ReadGeomSubsets(Scene scene, string path)
...
                 {
+                    var usdIndices = usdMesh.faceVertexIndices;
                     unityMesh.subMeshCount = geomSubsets.Subsets.Count;
-                    int subsetIndex = 0;
+                    var subsetIndex = 0;
                     foreach (var kvp in geomSubsets.Subsets)
                     {
-                        int[] faceIndices = kvp.Value;
-                        int[] triangleIndices = new int[faceIndices.Length * 3];
+                        var faceIndices = kvp.Value;
```
Comment (jcowles) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs]: Changing this to `var` makes the code harder to read, because we also don't know what types `kvp` holds.

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+                        newValues.Add(values[next]);
+                    }
+
+                last += faceVertexCount[i];
+            }
+
+            values = newValues.ToArray();
+        }
+
+        void ConvertInterpolationToFaceVarying<T>(ref T[] values, int[] vertexIndices, bool changeHandedness = false, TfToken interpolation = null)
```
Comment (jcowles) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs:372]: I think it's worth adding a doc string for this function, since converting TO FaceVarying for import sounds strange.

I would expect things to be converted to Vertex/Varying interpolation for import.

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+                Flatten(ref values, faceVertexIndices);
+            }
+            else if (interpolation == UsdGeomTokens.faceVarying) // 1 value per vertex per triangle
+            {
+                TriangulateAttributes(ref values, originalFaceVertexCounts, changeHandedness);
+            }
+        }
+
+        /// <summary>
+        /// Returns the interpolation of an data array based on the number of elements
```
Comment (jcowles) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs:399]: It's worth noting that this method, in general, will not always be correct.

It can be correct for constant and uniform interpolation, but varying/vertex/facevarying will sometimes be ambiguous.

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+using System.Runtime.CompilerServices;
+using pxr;
+using USD.NET;
+using USD.NET.Unity;
+using UnityEngine;
+
+[assembly: InternalsVisibleToAttribute("Unity.Formats.USD.Tests")]
+namespace Unity.Formats.USD
+{
+    public interface ISanitizable
```
Comment (jcowles) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs:12]: I love this idea - it awesomely simplifies import code and isolates the transformation into concise operators

**Review by jcowles — APPROVED** [review]
Wow, this change is incredible :)

It's clear that a massive amount of effort went into this, really nice work!

I have a few minor comments, but overall it totally makes sense and is a great simplification.

**Review by cguer — APPROVED** [review]
QA approved 👍!

Tested on:
- 2021.2.0b10 (Windows + HDRP)

Sorry for the delay 🙈 ...

File: package/com.unity.formats.usd/Dependencies/USD.NET.Unity/Geometry/MeshSampleBase.cs (hunk 1 above)
Comment (judubu) [reply package/com.unity.formats.usd/Dependencies/USD.NET.Unity/Geometry/MeshSampleBase.cs:39]: Because I could not find any reference to TexCoord4f. What are UV coords of type float4 by the way?

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs (hunk 6 above)
Comment (judubu) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs:399]: You're right, indexed arrays could easily mess it up and they are probably frequentely used.  
Indexed Vertex/Varying arrays will be shorter that points.length and indexed FaceVarying arrays will be shorter than faceVertexIndices.length or even equal to points.length in a worst case scenario!
All those situation will cause the MeshImporter to fail creating a valid mesh.

Comment (judubu) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs:399]: @jcowles Now the issue to solve is that we don't deserialize the interpolation type for normals and tangents (we don't do the indices either). Although we do it for the colors and uvs by defining them as Primvars. Normals and tangents are not technically USD primvars but do you see an issue if I set them as Primvars?

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs (hunk 5 above)
Comment (judubu) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs:372]: Added in dd63d81

File: package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs (hunk 4 above)
Comment (judubu) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs]: Added in dd63d81

File: package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/CubeImporter.cs (hunk 3 above)
Comment (judubu) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/CubeImporter.cs]: Fixed in dd63d81

File: package/com.unity.formats.usd/Dependencies/USD.NET/serialization/Primvar.cs (hunk 2 above)
Comment (judubu) [reply package/com.unity.formats.usd/Dependencies/USD.NET/serialization/Primvar.cs]: Fixed in dd63d81

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs (hunk 6 above)
Comment (judubu) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs:399]: Rereading the doc I had some of that wrong:
- normals can be "builtin" or primvars
- only primvar normals can be indexed
- builtin normals are subject to interpolation but not indexing

Primvar normals should normally take precedence on builtin normals but we dont support them at the moment. It leaves us with non indexed builtin normals for which the GuessInterpolation method should work.

@jcowles  What is the ambiguous situation you had in mind?

File: package/com.unity.formats.usd/Dependencies/USD.NET.Unity/Geometry/MeshSampleBase.cs (hunk 1 above)
Comment (jcowles) [reply package/com.unity.formats.usd/Dependencies/USD.NET.Unity/Geometry/MeshSampleBase.cs:39]: Unity has float4 uvs, which are useful sometimes for storing tangent data, etc

File: package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs (hunk 6 above)
Comment (jcowles) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs:399]: Indexed primvars for sure, but there are degenerate non-indexed cases, where you have four faces and all faces share the same verts - uniform, vertex/varying, face varying are not distinguishable. Or more common cases, e.g. if you have a single face, then vertex & face-varying are not distinguishable.

I think what you have is fine, but it's worth calling this out.

Other primitives (such as curves) have even more aliasing.

Comment (judubu) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs:399]: Got it thanks for the update. Adding that as comment in the code.
